### PR TITLE
Feature/refresh token

### PR DESCRIPTION
**Pull request by eliasrabelomatos**

### CODE REVIEW

File: src/repositories/accessToken.repository.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+const { AccessToken } = require('../models');
+
+module.exports = {
+  create: (args) => AccessToken.create(args),
+  get: (args) => AccessToken.findOne(args),
+  update: (obj, id) => AccessToken.update(obj, { where: { id } }),
```
Comment (lucas-a-pelegrino) [inline src/repositories/accessToken.repository.js:6]: Siga o padrão do model de Users pro update. O `.save` nesse caso é melhor.

File: src/services/accessToken/create.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+module.exports = {
+  create: async (params) => {
+    const token = encryptor.generateToken(params, {
+      algorithm: 'HS384',
+      expiresIn: 86000,
+    });
+    const refreshParams = { id: params.sub.id };
+    const refreshToken = encryptor.generateToken(refreshParams, {
+      algorithm: 'HS256',
+      expiresIn: 86000 * 30,
```
Comment (lucas-a-pelegrino) [inline src/services/accessToken/create.js:13]: O ideal é que esse tempo de expiração fique no `.env`.

File: src/services/accessToken/refreshToken.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+const httpCodes = require('http-status-codes');
```
Comment (lucas-a-pelegrino) [inline src/services/accessToken/refreshToken.js:1]: Deixe esse arquivo junto com os services do **auth**, pois ele faz parte do fluxo de autenticação/autorização.

File: src/services/accessToken/create.service.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+const { accessTokenRepository } = require('../../repositories');
```
Comment (lucas-a-pelegrino) [inline src/services/accessToken/create.service.js:1]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/config/database/migrations/20200802012857-create-access-tokens.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        references: { model: 'users', key: 'id' },
+        onUpdate: 'CASCADE',
+        onDelete: 'CASCADE',
+        allowNull: false,
+      },
+      token: {
+        type: Sequelize.TEXT,
+        allowNull: false,
+      },
+      refresh_token: {
```
Comment (lucas-a-pelegrino) [inline src/config/database/migrations/20200802012857-create-access-tokens.js:21]: Codacy found an issue: [Identifier 'refresh_token' is not in camel case.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/auth/refreshToken.service.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+const httpCodes = require('http-status-codes');
+const moment = require('moment');
```
Comment (lucas-a-pelegrino) [inline src/services/auth/refreshToken.service.js:2]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/config/database/migrations/20200802012857-create-access-tokens.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.createTable('access_tokens', {
+      id: {
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      user_id: {
```
Comment (lucas-a-pelegrino) [inline src/config/database/migrations/20200802012857-create-access-tokens.js:10]: Codacy found an issue: [Identifier 'user_id' is not in camel case.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/auth/signin.service.js [hunk 8]
```diff
@@ -3,6 +3,7 @@ const moment = require('moment');
 const { usersRepository } = require('../../repositories');
 const { ApplicationError } = require('../../utils');
 const { encryptor } = require('../../helpers');
+const accessTokenService = require('../accessToken');
```
Comment (lucas-a-pelegrino) [inline src/services/auth/signin.service.js:6]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/models/AccessToken.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+module.exports = (sequelize, DataTypes) => {
+  const AccessToken = sequelize.define(
+    'AccessToken',
+    {
+      userId: {
+        type: DataTypes.INTEGER,
+        field: 'user_id',
```
Comment (lucas-a-pelegrino) [inline src/models/AccessToken.js:7]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/models/AccessToken.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        field: 'refresh_token',
+      },
+      expired: DataTypes.BOOLEAN,
+      createdAt: {
+        type: DataTypes.DATE,
+        field: 'created_at',
+      },
+      updatedAt: {
+        type: DataTypes.DATE,
+        field: 'updated_at',
```
Comment (lucas-a-pelegrino) [inline src/models/AccessToken.js:21]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/auth/refreshToken.service.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+const { Op } = require('sequelize');
+const { jwt, ApplicationError } = require('../../utils');
```
Comment (lucas-a-pelegrino) [inline src/services/auth/refreshToken.service.js:4]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/config/database/migrations/20200802012857-create-access-tokens.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        type: Sequelize.INTEGER,
+        references: { model: 'users', key: 'id' },
```
Comment (lucas-a-pelegrino) [inline src/config/database/migrations/20200802012857-create-access-tokens.js:12]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/models/AccessToken.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+      },
+    },
+    { tableName: 'access_tokens' },
```
Comment (lucas-a-pelegrino) [inline src/models/AccessToken.js:24]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/models/AccessToken.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+  );
+  AccessToken.associate = function associate(models) {
+    models.AccessToken.belongsTo(models.User, {
+      foreignKey: 'userId',
```
Comment (lucas-a-pelegrino) [inline src/models/AccessToken.js:28]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/config/database/migrations/20200802012857-create-access-tokens.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.createTable('access_tokens', {
```
Comment (lucas-a-pelegrino) [inline src/config/database/migrations/20200802012857-create-access-tokens.js:3]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/accessToken/create.service.js [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+const { encryptor } = require('../../helpers');
+
+module.exports = {
+  create: async (params) => {
+    const token = encryptor.generateToken(params, {
+      algorithm: 'HS384',
```
Comment (lucas-a-pelegrino) [inline src/services/accessToken/create.service.js:7]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/accessToken/create.service.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+      expiresIn: process.env.ACCESS_TOKEN_EXPIRES_IN,
+    });
+    const refreshParams = { id: params.sub.id };
+    const refreshToken = encryptor.generateToken(refreshParams, {
+      algorithm: 'HS256',
```
Comment (lucas-a-pelegrino) [inline src/services/accessToken/create.service.js:12]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/models/AccessToken.js [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+      },
+      token: DataTypes.STRING,
+      refreshToken: {
+        type: DataTypes.STRING,
+        field: 'refresh_token',
+      },
+      expired: DataTypes.BOOLEAN,
+      createdAt: {
+        type: DataTypes.DATE,
+        field: 'created_at',
```
Comment (lucas-a-pelegrino) [inline src/models/AccessToken.js:17]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/middlewares/isAuthorized.js [hunk 19]
```diff
@@ -1,5 +1,6 @@
+const { Op } = require('sequelize');
```
Comment (lucas-a-pelegrino) [inline src/middlewares/isAuthorized.js:1]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/repositories/accessToken.repository.js [hunk 20]
```diff
@@ -0,0 +1,7 @@
+const { AccessToken } = require('../models');
```
Comment (lucas-a-pelegrino) [inline src/repositories/accessToken.repository.js:1]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/accessToken/index.js [hunk 21]
```diff
@@ -0,0 +1,5 @@
+const { create } = require('./create.service');
```
Comment (lucas-a-pelegrino) [inline src/services/accessToken/index.js:1]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/index.js [hunk 22]
```diff
@@ -1,7 +1,9 @@
 const usersService = require('./users');
 const authService = require('./auth');
+const accessTokenService = require('./accessToken');
```
Comment (lucas-a-pelegrino) [inline src/services/index.js:3]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/auth/refreshToken.service.js [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+const { create } = require('../accessToken/create.service');
```
Comment (lucas-a-pelegrino) [inline src/services/auth/refreshToken.service.js:5]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/config/database/migrations/20200802012857-create-access-tokens.js [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+      id: {
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      user_id: {
+        type: Sequelize.INTEGER,
+        references: { model: 'users', key: 'id' },
+        onUpdate: 'CASCADE',
```
Comment (lucas-a-pelegrino) [inline src/config/database/migrations/20200802012857-create-access-tokens.js:13]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/config/database/migrations/20200802012857-create-access-tokens.js [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+        type: Sequelize.DATE,
+      },
+      updated_at: {
+        allowNull: false,
+        type: Sequelize.DATE,
+      },
+    });
+  },
+  down: (queryInterface) => {
+    return queryInterface.dropTable('access_tokens');
```
Comment (lucas-a-pelegrino) [inline src/config/database/migrations/20200802012857-create-access-tokens.js:41]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/config/database/migrations/20200802012857-create-access-tokens.js [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+      expired: {
+        type: Sequelize.BOOLEAN,
+        allowNull: false,
+        defaultValue: false,
+      },
+      created_at: {
+        allowNull: false,
+        type: Sequelize.DATE,
+      },
+      updated_at: {
```
Comment (lucas-a-pelegrino) [inline src/config/database/migrations/20200802012857-create-access-tokens.js:34]: Codacy found an issue: [Identifier 'updated_at' is not in camel case.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/auth/refreshToken.service.js [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+        [Op.and]: [
+          { token, refreshToken },
+          {
+            expired: false,
+          },
+        ],
+      },
+    });
+    if (!accessToken) {
+      throw new ApplicationError('token-not-found', httpCodes.NOT_FOUND);
```
Comment (lucas-a-pelegrino) [inline src/services/auth/refreshToken.service.js:28]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/models/AccessToken.js [hunk 28]
```diff
@@ -0,0 +1,32 @@
+module.exports = (sequelize, DataTypes) => {
+  const AccessToken = sequelize.define(
+    'AccessToken',
```
Comment (lucas-a-pelegrino) [inline src/models/AccessToken.js:3]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393)

File: src/services/accessToken/create.service.js [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+      expiresIn: process.env.REFRESH_TOKEN_EXPIRES_IN,
```
Comment (lucas-a-pelegrino) [inline src/services/accessToken/create.service.js:13]: Mesma observação de cima.

**Review by lucas-a-pelegrino — DISMISSED** [review]

**Comment by lucas-a-pelegrino** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/models/AccessToken.js  1
- src/config/database/migrations/20200802012857-create-access-tokens.js  1
- src/repositories/accessToken.repository.js  1
- src/services/auth/refreshToken.service.js  2
- src/services/accessToken/create.service.js  1
         

Complexity decreasing per file
==============================
+ src/controllers/auth.controller.js  -1
+ src/middlewares/isAuthorized.js  -3
         

Clones added
============
- src/config/database/migrations/20200802012857-create-access-tokens.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393&bid=19504189)

**Comment by lucas-a-pelegrino** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/services/auth/refreshToken.service.js  2
- src/repositories/accessToken.repository.js  1
- src/services/accessToken/create.service.js  1
- src/models/AccessToken.js  1
- src/config/database/migrations/20200802012857-create-access-tokens.js  1
         

Complexity decreasing per file
==============================
+ src/middlewares/isAuthorized.js  -3
+ src/controllers/auth.controller.js  -1
         

Clones added
============
- src/config/database/migrations/20200802012857-create-access-tokens.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393&bid=19504189)

**Comment by lucas-a-pelegrino** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/config/database/migrations/20200802012857-create-access-tokens.js  1
- src/services/accessToken/create.service.js  1
- src/repositories/accessToken.repository.js  1
- src/models/AccessToken.js  1
- src/services/auth/refreshToken.service.js  2
         

Complexity decreasing per file
==============================
+ src/middlewares/isAuthorized.js  -3
+ src/controllers/auth.controller.js  -1
         

Clones added
============
- src/config/database/migrations/20200802012857-create-access-tokens.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393&bid=19504189)

**Comment by lucas-a-pelegrino** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/services/auth/refreshToken.service.js  2
- src/repositories/accessToken.repository.js  1
- src/services/accessToken/create.service.js  1
- src/models/AccessToken.js  1
         

Complexity decreasing per file
==============================
+ src/controllers/auth.controller.js  -1
+ src/middlewares/isAuthorized.js  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393&bid=19504189)

**Review by diogo26 — APPROVED** [review]

**Comment by lucas-a-pelegrino** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/repositories/accessToken.repository.js  1
- src/services/accessToken/create.service.js  1
- src/models/AccessToken.js  1
- src/services/auth/refreshToken.service.js  2
         

Complexity decreasing per file
==============================
+ src/middlewares/isAuthorized.js  -3
+ src/controllers/auth.controller.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lucas.assuncao.p/node-bloodboiler-sequelized/pullRequest?prid=5950393&bid=19504189)

**Pull request merged**